crates/extensions/c8y_mapper_ext/src/converter.rs

```diff
@@ -1443,10 +1450,20 @@ impl CumulocityConverter {
 
     async fn register_software_list_operation(
         &self,
-        _target: &EntityTopicId,
+        target: &EntityTopicId,
+        message: &Message,
     ) -> Result<Vec<Message>, ConversionError> {
-        // On c8y, "software list" is implied by "software update"
-        Ok(vec![])
+        if !self.config.software_management_with_types {
+            debug!("Publishing c8y_SupportedSoftwareTypes is disabled. To enable it, run `tedge config set c8y.software_management.with_types true`.");
+            return Ok(vec![]);
+        }
+
+        // Send c8y_SupportedSoftwareTypes, which is introduced in c8y >= 10.14
+        let data = SoftwareCommandMetadata::from_json(message.payload_str()?)?;
+        let payload = json!({"c8y_SupportedSoftwareTypes": data.types}).to_string();
+        let topic = self.get_inventory_update_topic(target)?;
```

Comment:
This PR is not the right place to improve on this, but we will have to find a better pattern before being overwhelmed by this kind of artificial and error-prone complexity.

- Artificial, because we only need to translate a tedge MQTT message into a c8y one. This is not obvious from these lines.
- Error-prone, because the message must be partially parsed before calling this method (to extract the target and message topic from the message.topic).
- I agree this is not so easy, as we need to access the inventory to translate the tedge entity id into a c8y external id.

Reply:
Yeah we can work to improve, however, I'm not sure how much I should do in this PR...

By the way, since this action updates the device's c8y inventory, should the mapper publish to our `twin` topic, instead of directly publishing to the JSON over MQTT topic?

Reply:
> I'm not sure how much I should do in this PR...

Definitely not in this PR as this would impact all the messages received by the c8y converter.

Reply:
> should the mapper publish to our twin topic, instead of directly publishing to the JSON over MQTT topic?

No, the `twin` topic is where child devices publish their metadata. The c8y mapper should react on these messages, not push data there.